### PR TITLE
Add k0scontrolplane heathcheck-remediation

### PR DESCRIPTION
**Pull request by apedriza**
This PR adds the reconciliation by k0scontrolplane of machines that are considered unhealthy by the machinehealtcheck controller. Check [MachineHealthCheck contract](https://cluster-api.sigs.k8s.io/tasks/automated-machine-management/healthchecking) for more details h

It basically replicates the behavior of KubeadmControlPlane when handling machines considered unhealthy except that a remediation strategy in order to have a more granular process control is not implemented. Currently machine creation does not take into account the previous state of a machine if it is to be a replacement, so adding this control would require changes to the machine synchronization process. It can always be added later but I did not want to compromise that logic in this PR given its sensitivity.

### CODE REVIEW

File: internal/controller/controlplane/remediation.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+func (c *K0sController) reconcileUnhealthyMachines(ctx context.Context, cluster *clusterv1.Cluster, kcp *cpv1beta1.K0sControlPlane) (retErr error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	machines, err := collections.GetFilteredMachinesForCluster(ctx, c, cluster, collections.ControlPlaneMachines(cluster.Name))
+	if err != nil {
+		return fmt.Errorf("failed to filter machines for control plane: %w", err)
+	}
+
+	healthyMachines := machines.Filter(collections.Not(isUnhealthy))
```
Comment (makhov) [inline internal/controller/controlplane/remediation.go:42]: Could we use `collections.Not(collections.HasUnhealthyCondition)` here? If not, could we then avoid a double negative here? Eg something like `machines.Filter(isHealthy)`?

Comment (apedriza) [reply internal/controller/controlplane/remediation.go]: 👍 fixed using `machines.Filter(isHealthy)`

**Review by makhov — COMMENTED** [review]
Sorry, just noticed a couple of things with annotations.

File: internal/controller/controlplane/helper.go [hunk 2]
```diff
@@ -52,9 +52,18 @@ func (c *K0sController) createMachine(ctx context.Context, name string, cluster
...
+	err = c.Client.Patch(ctx, machine, client.Apply, &client.PatchOptions{
 		FieldManager: "k0smotron",
 	})
+	if err != nil {
+		return machine, err
+	}
+
+	// Remove the annotation tracking that a remediation is in progress.
+	// A remediation is completed when the replacement machine has been created above.
+	delete(kcp.Annotations, cpv1beta1.RemediationInProgressAnnotation)
```
[the remark this answers is not in the document]
Comment (apedriza) [reply internal/controller/controlplane/helper.go:64]: in this way we are sure machine is created in kube-api which means any remediation is done. If not there could be errors creating machine in kube-api and start a second remediation even if the first one was not completed. I think is safer if we make sure machine is created/remediated. WDYT?

File: internal/controller/controlplane/remediation.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+	if err := c.runMachineDeletionSequence(ctx, log, cluster, kcp, machineToBeRemediated); err != nil {
+		conditions.MarkFalse(machineToBeRemediated, clusterv1.MachineOwnerRemediatedCondition, clusterv1.RemediationFailedReason, clusterv1.ConditionSeverityError, err.Error())
+		return errors.Wrapf(err, "failed to delete unhealthy machine %s", machineToBeRemediated.Name)
+	}
+	log.Info("Remediated unhealthy machine, another new machine should take its place soon.")
+
+	// Mark controlplane to track that remediation is in progress and do not proceed until machine is gone.
+	// This annotation is removed when new controlplane creates a new machine.
+	annotations.AddAnnotations(kcp, map[string]string{
```
[the remark this answers is not in the document]
Comment (apedriza) [reply internal/controller/controlplane/remediation.go:120]: I think removing it before creating the machine is safe in order to continue with next remediations. We could face cases where more than one machine needs to be remediated. This annotation is to not allow multiples remediations at the same time

**Comment by makhov** [comment]
@AdrianPedriza looks like something is off with changes in this PR currently

**Review by makhov — APPROVED** [review]

**Pull request merged**